File: sdks/python/examples/07_frame_metadata.py
#!/usr/bin/env python3
"""07 · Frame metadata — what arrives alongside the pixels.

Every decoded frame can carry a trailer the sender attached: a frame id, a
wall-clock timestamp, and arbitrary bytes of its own. `on_raw_frame` hands all of
it over; a frame with no trailer arrives with zeros and empty bytes instead.

    uv run python examples/07_frame_metadata.py --seconds 10

The timestamps are what makes this more than trivia: they are the sender's clock,
so they measure the model's own pacing rather than the network's — two frames
that arrive together were not necessarily produced together.

`user_data` is empty unless the model puts it there: a model that derives its
output from your frame can mirror the bytes back, and none of the published ones
does today. 04 shows the sending side of the same field.

Native-only: a browser hands JS a MediaStreamTrack with no per-frame hook, so
there is no JS counterpart to this example.

Docs:
  Frame metadata  https://docs.reactor.inc/concepts/frame-metadata
  Track (API)     https://docs.reactor.inc/sdk-reference/python/track
"""

from __future__ import annotations

import asyncio

import common

from reactor_sdk import Reactor


async def main() -> None:
    args = common.parse(__doc__)

    reactor = Reactor(
        model_name=args.model,
        api_key=args.api_key,
        jwt=args.jwt,
        api_url=args.api_url,
        local=args.local,
    )

    @reactor.on_status
    def status(new: str) -> None:
        print(f"status: {new}")

    @reactor.on_error
    def failed(error: Exception) -> None:
        print(f"error: {error}")

    async with reactor:
        await reactor.connect()
        print(f"session: {reactor.session_id}")
        await common.bootstrap(reactor, args.model)

        output = reactor.tracks.with_kind("video").with_direction("recvonly").one()
        print(f"track: {output.name}")

        window = common.display(args, f"{args.model} · {output.name}")

        with_trailer = 0
        without_trailer = 0
        tagged = 0
        previous_timestamp: int | None = None
        gaps: list[int] = []

        @output.on_raw_frame
        def inspect(
            data: bytes,
            width: int,
            height: int,
            frame_id: int,
            timestamp_us: int,
            user_data: bytes,
        ) -> None:
            nonlocal with_trailer, without_trailer, tagged, previous_timestamp
            window.submit(data, width, height)
            # No trailer at all looks like this: the ids and the timestamp are
            # zero and there are no bytes. Worth checking before trusting either.
            if not frame_id and not timestamp_us and not user_data:
                without_trailer += 1
                return

            with_trailer += 1
            if user_data:
                tagged += 1
            if previous_timestamp is not None and timestamp_us > previous_timestamp:
                gaps.append(timestamp_us - previous_timestamp)
            previous_timestamp = timestamp_us

            if with_trailer <= 3:
                print(
                    f"frame {frame_id}: {width}x{height} "
                    f"at {timestamp_us} us, user_data={user_data!r}"
                )

        await window.hold(args.seconds)

    print(f"frames with a trailer: {with_trailer}")
    print(f"frames without one: {without_trailer}")
    print(f"frames carrying user_data: {tagged}")
    if gaps:
        ordered = sorted(gaps)
        median_us = ordered[len(ordered) // 2]
        print(f"sender cadence: median {median_us / 1000:.1f} ms between frames")
        print(f"                {1_000_000 / median_us:.1f} fps as the sender timed it")


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("interrupted")

File: sdks/cpp/examples/07_frame_metadata.cpp
// 07 — The trailer on each incoming frame.
//
// Every frame can carry three things beyond its pixels: the sender's frame id,
// the sender's capture time, and whatever bytes the sender tagged it with. All
// three are optional, and a frame without them arrives as zeros and no bytes.
//
// **Expect zeros against Helios.** This example's target model is purely
// generative — it has no input frames at all, so there is nothing of the
// client's for it to mirror back, and a trailer here reads
// `frame_id=false timestamp=false user_data=false` by construction, not
// because no model bothers. `reactor/echo` is different: as of version 1.7.5
// it reads a client's own webcam frames and mirrors whatever `user_data` was
// attached to them straight onto `main_video`. What this example is for
// either way is the *reading* side. Example 04 shows the sending side.
//
// The capture time is in the *engine's* clock, the one `reactor::time_micros()`
// reads. It is not a UNIX timestamp and comparing it to one is meaningless; what
// it is good for is the difference between two frames, and the offset from a local
// read at the same moment.
//
//   export REACTOR_API_KEY=...
//   ./07_frame_metadata
//
// Docs: https://docs.reactor.inc/concepts/frame-metadata

#include <algorithm>
#include <chrono>
#include <cstdlib>
#include <iostream>
#include <mutex>
#include <reactor/reactor.hpp>
#include <string>
#include <thread>
#include <vector>

namespace {

constexpr auto MODEL = "reactor/helios";

struct Trailer {
  std::uint64_t frame_id = 0;
  std::uint64_t timestamp_us = 0;
  std::size_t user_data_bytes = 0;
  std::int64_t local_time_us = 0;
};

}  // namespace

int main() {
  const char* api_key = std::getenv("REACTOR_API_KEY");
  if (api_key == nullptr) {
    std::cerr << "set REACTOR_API_KEY\n";
    return 2;
  }

  reactor::Reactor client{MODEL, reactor::ApiKey{api_key}};
  auto errors = client.on_error(
      [](const reactor::ReactorError& error) { std::cerr << "error: " << error.what() << '\n'; });

  try {
    client.connect().get();
    client.send_command("set_prompt", {{"prompt", "a hummingbird at a feeder, slow motion"}}).get();
    client.send_command("start").get();

    std::mutex mutex;
    std::vector<Trailer> seen;

    auto subscription = client.track("main_video").on_frame([&](const reactor::VideoFrame& frame) {
      Trailer trailer;
      trailer.frame_id = frame.frame_id;
      trailer.timestamp_us = frame.timestamp_us;
      trailer.user_data_bytes = frame.user_data.size;
      // Read on this side, at the moment of arrival, in the same clock the
      // sender stamped with. The difference is the only thing either value
      // means on its own.
      trailer.local_time_us = reactor::time_micros();

      const std::lock_guard<std::mutex> lock(mutex);
      seen.push_back(trailer);
    });

    std::cout << "collecting frames for 10s\n";
    std::this_thread::sleep_for(std::chrono::seconds(10));

    const std::lock_guard<std::mutex> lock(mutex);
    std::cout << "received " << seen.size() << " frames\n";
    if (seen.empty()) {
      client.disconnect().get();
      return 1;
    }

    const bool has_ids =
        std::any_of(seen.begin(), seen.end(), [](const Trailer& t) { return t.frame_id != 0; });
    const bool has_times =
        std::any_of(seen.begin(), seen.end(), [](const Trailer& t) { return t.timestamp_us != 0; });
    const bool has_tags = std::any_of(seen.begin(), seen.end(),
                                      [](const Trailer& t) { return t.user_data_bytes != 0; });

    std::cout << "trailer present: frame_id=" << std::boolalpha << has_ids
              << " timestamp=" << has_times << " user_data=" << has_tags << '\n';
    if (!has_ids && !has_times && !has_tags) {
      std::cout << "  (all zero — expected against Helios, which has no input frames "
                   "to mirror; see the note at the top)\n";
    }

    for (std::size_t index = 0; index < std::min<std::size_t>(seen.size(), 5); ++index) {
      const auto& trailer = seen[index];
      std::cout << "  frame_id=" << trailer.frame_id << " timestamp_us=" << trailer.timestamp_us;
      if (trailer.timestamp_us != 0) {
        // Positive: the frame was captured before it arrived here. Large means the
        // model is generating slower than real time, which is normal and is why
        // clip readiness is measured in media time.
        std::cout << " age_ms="
                  << (trailer.local_time_us - static_cast<std::int64_t>(trailer.timestamp_us)) /
                         1000;
      }
      std::cout << " user_data=" << trailer.user_data_bytes << "B\n";
    }

    if (has_ids && seen.size() > 1) {
      // Contiguous ids mean nothing was dropped between the sender and here. A gap
      // is not an error: the FFI keeps the newest frame while a handler runs.
      std::size_t gaps = 0;
      for (std::size_t index = 1; index < seen.size(); ++index) {
        if (seen[index].frame_id != seen[index - 1].frame_id + 1) {
          ++gaps;
        }
      }
      std::cout << "gaps in the frame_id sequence: " << gaps << '\n';
    }

    client.disconnect().get();
    return 0;
  } catch (const reactor::ReactorError& error) {
    std::cerr << "failed: " << error.what() << '\n';
    try {
      client.disconnect().get();
    } catch (const reactor::ReactorError&) {
    }
    return 1;
  }
}
